File: Lista_Ligada.py
class No:
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None

class ListaLigada:
    def __init__(self):
        self.cabeca = None

    def inserir_inicio(self, valor):
        novo = No(valor)
        novo.proximo = self.cabeca
        self.cabeca = novo

    def inserir_fim(self, valor):
        novo = No(valor)
        if not self.cabeca:
            self.cabeca = novo
            return
        atual = self.cabeca
        while atual.proximo:
            atual = atual.proximo
        atual.proximo = novo

    def remover(self, valor):
        if not self.cabeca:
            return False
        if self.cabeca.valor == valor:
            self.cabeca = self.cabeca.proximo
            return True
        atual = self.cabeca
        while atual.proximo and atual.proximo.valor != valor:
            atual = atual.proximo
        if atual.proximo:
            atual.proximo = atual.proximo.proximo
            return True
        return False

    def buscar(self, valor):
        atual = self.cabeca
        while atual:
            if atual.valor == valor:
                return True
            atual = atual.proximo
        return False

    def exibir(self):
        valores = []
        atual = self.cabeca
        while atual:
            valores.append(str(atual.valor))
            atual = atual.proximo
        print(" -> ".join(valores) + " -> None")

# Teste
if __name__ == "__main__":
    lista = ListaLigada()
    lista.inserir_fim(10)
    lista.inserir_fim(20)
    lista.inserir_inicio(5)
    lista.exibir()  # 5 -> 10 -> 20 -> None
    print("Buscar 10:", lista.buscar(10))
    lista.remover(10)
    lista.exibir()

File: ListaLigada.cs
using System;

public class No
{
    public int Valor;
    public No? Proximo;
    public No(int valor) { Valor = valor; Proximo = null; }
}

public class ListaLigada
{
    private No? cabeca;

    public void InserirInicio(int valor)
    {
        No novo = new No(valor);
        novo.Proximo = cabeca;
        cabeca = novo;
    }

    public void InserirFim(int valor)
    {
        No novo = new No(valor);
        if (cabeca == null) { cabeca = novo; return; }
        No atual = cabeca;
        while (atual.Proximo != null) atual = atual.Proximo;
        atual.Proximo = novo;
    }

    public bool Remover(int valor)
    {
        if (cabeca == null) return false;
        if (cabeca.Valor == valor) { cabeca = cabeca.Proximo; return true; }
        No atual = cabeca;
        while (atual.Proximo != null && atual.Proximo.Valor != valor)
            atual = atual.Proximo;
        if (atual.Proximo != null)
        {
            atual.Proximo = atual.Proximo.Proximo;
            return true;
        }
        return false;
    }

    public bool Buscar(int valor)
    {
        No? atual = cabeca;
        while (atual != null)
        {
            if (atual.Valor == valor) return true;
            atual = atual.Proximo;
        }
        return false;
    }

    public void Exibir()
    {
        No? atual = cabeca;
        while (atual != null)
        {
            Console.Write($"{atual.Valor} -> ");
            atual = atual.Proximo;
        }
        Console.WriteLine("null");
    }
}

class Program
{
    static void Main()
    {
        var lista = new ListaLigada();
        lista.InserirFim(10);
        lista.InserirFim(20);
        lista.InserirInicio(5);
        lista.Exibir();
        Console.WriteLine($"Buscar 10: {lista.Buscar(10)}");
        lista.Remover(10);
        lista.Exibir();
    }
}
